src/Java.Interop/Java.Interop/JavaPrimitiveArrays.cs

```diff
@@ -11,28 +11,90 @@
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaPrimitiveArray<Char>),  new JniTypeSignature ("C", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaArray<Char>),           new JniTypeSignature ("C", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaPrimitiveArray<Int16>),  new JniTypeSignature ("S", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaArray<Int16>),           new JniTypeSignature ("S", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaPrimitiveArray<Int32>),  new JniTypeSignature ("I", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaArray<Int32>),           new JniTypeSignature ("I", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaPrimitiveArray<Int64>),  new JniTypeSignature ("J", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaArray<Int64>),           new JniTypeSignature ("J", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaPrimitiveArray<Single>),  new JniTypeSignature ("F", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaArray<Single>),           new JniTypeSignature ("F", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaPrimitiveArray<Double>),  new JniTypeSignature ("D", arrayRank: 1, keyword: true)),
-				new KeyValuePair<Type, JniTypeSignature>(typeof (JavaArray<Double>),           new JniTypeSignature ("D", arrayRank: 1, keyword: true)),
-			};
+		static JniTypeSignature __BooleanTypePrimitiveArraySignature;
+		static JniTypeSignature __BooleanTypeArraySignature;
+		static JniTypeSignature __SByteTypePrimitiveArraySignature;
+		static JniTypeSignature __SByteTypeArraySignature;
+		static JniTypeSignature __CharTypePrimitiveArraySignature;
+		static JniTypeSignature __CharTypeArraySignature;
+		static JniTypeSignature __Int16TypePrimitiveArraySignature;
+		static JniTypeSignature __Int16TypeArraySignature;
+		static JniTypeSignature __Int32TypePrimitiveArraySignature;
+		static JniTypeSignature __Int32TypeArraySignature;
+		static JniTypeSignature __Int64TypePrimitiveArraySignature;
+		static JniTypeSignature __Int64TypeArraySignature;
+		static JniTypeSignature __SingleTypePrimitiveArraySignature;
+		static JniTypeSignature __SingleTypeArraySignature;
+		static JniTypeSignature __DoubleTypePrimitiveArraySignature;
+		static JniTypeSignature __DoubleTypeArraySignature;
+
+		static bool GetBuiltInTypeArraySignature (Type type, ref JniTypeSignature signature)
+		{
+			if (type == typeof (JavaPrimitiveArray<Boolean>)) {
+				signature = GetCachedTypeSignature (ref __BooleanTypePrimitiveArraySignature, "Z", arrayRank: 1, keyword: true);
```

Comment:
Why *both* `__BooleanTypePrimitiveArraySignature` *and* `__BooleanTypeArraySignature`?  They both have the same signature, `[Z`, and would be Equal to each other.  Why not just have the single field?

Reply:
I didn't want to change the old code semantics, and it kept the two types distinct (maybe because it was all stored in an array) and I always aim to make the minimal set of changes necessary.